compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -530,7 +536,18 @@ fn get_argc_argv<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     cx: &'a Bx::CodegenCx,
     bx: &mut Bx,
 ) -> (Bx::Value, Bx::Value) {
-    if cx.sess().target.main_needs_argc_argv {
+    if cx.sess().target.os.contains("uefi") {
+        // Params for UEFI
+        let param_handle = bx.get_param(0);
+        let param_system_table = bx.get_param(1);
+        let arg_argc = bx.const_int(cx.type_isize(), 2);
+        let arg_argv = bx.alloca(cx.type_array(cx.type_i8p(), 2), Align::ONE);
+        bx.store(param_handle, arg_argv, Align::ONE);
+        let arg_argv_el1 =
+            bx.gep(cx.type_ptr_to(cx.type_i8()), arg_argv, &[bx.const_int(cx.type_int(), 1)]);
+        bx.store(param_system_table, arg_argv_el1, Align::ONE);
+        (arg_argc, arg_argv)
```

Comment:
Could you clarify what the concern is here, exactly? Are we worrying about conflicting with someone else trying to use argc and argv, basically?

Reply:
I don't think there is an inherent conflict. But we would effectively make it impossible to use `#[start]` without platform knowledge. Because any crate that uses `#[start]` must be adjusted to `NOT` assume `argv` to be a c-string array for some platforms.

Arguments passed to the individual startup routines differ strongly across platforms. Up until now `#[start]` simply got argc+argv in the unix-style *iff* the platform passed it, and "0/[0; 1]" (or "0/NULL", not sure) otherwise. Other arguments were just silently dropped, since the platform provides separate accessors (maybe I am wrong and there is precedence; but I didn't see any so far).

There is some discussion in the tracking issue of `#[start]` (https://github.com/rust-lang/rust/issues/29633), and I feel like there is consensus in using a custom enum to pass the data. But this is not like anyone is working on it, or any urgent need to implement it.

Reply:
I believe complicating the use of an unstable feature which is, as you noted, fairly "stuck", is undesirable but not unexpected. I do not believe a slightly hacky implementation on _this_ count is sufficient reason to bar this PR, although it ought to come with an appropriately huge `// FIXME`, _unless_ you believe that this choice will somehow become impossible to overcome later.